crypto/hmac/hmac.c

```diff
@@ -17,10 +17,34 @@
 #include <stdlib.h>
 #include <string.h>
 #include "internal/cryptlib.h"
-#include <openssl/hmac.h>
 #include <openssl/opensslconf.h>
+#include <openssl/hmac.h>
+#include <openssl/core_names.h>
 #include "hmac_local.h"
 
+unsigned char *OSSL_HMAC(OSSL_LIB_CTX *libctx, const char *propq,
```

Comment:
With `EVP_Q_mac`, this should be unnecessary...  and quite frankly, I don't see why we should add back algorithm specific functions when it's fairly easy to make generic ones with similar enough functionality.

Reply:
For the sake of usability I strongl disagree here. 
What you call "fairly easy" in practice looks like this at best:
```
    OSSL_PARAM params[2] =
        { OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
                                           (char *)digest, 0),
          OSSL_PARAM_END };

    return OSSL_Q_mac(libctx, "HMAC", propq, params, key, keylen,
                      data, datalen, out, outsize, outlen);
```

I do not want to impose the burden on lib users (including myself, but also part of the crypto lib itself and the OpenSSL apps) to set up a parameter array like this for each place in the code where an HMAC calculation is need.
(Flexible parameters would be less cumbersome for instance in Perl, but we are using C here.)

Reply:
... and when the next great MAC algo comes along (let's call i BLARGH), with a provider and all?  ...  or are you reserving these names only for algos that have been implemented up until now, leaving the rest for the more general function?

Reply:
> ... and when the next great MAC algo comes along (let's call i BLARGH), 

When BLARGH has come along, users must decide whether it's worth adapting existing code.
Yet at this point our discrepancy is not actually about backward compatibility but about plain ease of use.
So in this example, I would advise for providing an API for BLARGH that should be as easy to use as currently for HMAC.

(And in case the decision is to make use of the new algo, one would just need to replace the rather simple call of `OSSL_HMAC(...)` by `OSSL_BLARGH(...)`.)

Reply:
If that becomes and expectation, then we're making ourselves target for pressure to make certain releases *based on release of other projects*, notably those that produce providers for new algos.

Reply:
I consider the issues with using new algos (which anyway will take time to get adopted)
less pressing than the issues with adapting to a new crypto lib version, which usually is required to do rather timely for security and SW maintenance reasons.

In this sub-thread the topic the usability of current algo implementations (here, HMAC).
Whenever a new (e.g., BLARGH) algo implementation/provider is privided by whoever, it won't be hard and pretty quick to add a convenience function for that new BLARGH algo with the next OpenSSL minor release.
And if part of the users are extremely eager to use the new BLARGH algo before that, they can safely use the most flexible (though more cumbersome to use) general OpenSSL API.

Reply:
> What you call "fairly easy" in practice looks like this at best:

I wonder how difficult it would be to put in place a simple string->OSSL_PARAM conversion function. E.g. the user calls:

````
OSSL_Q_mac(libctx, "HMAC", propq, "digest=SHA256", key, keylen,
                      data, datalen, out, outsize, outlen);
````

Internally,`OSSL_Q_mac` would parse "digest=SHA256" into a key/name pair and convert it to:
````
    OSSL_PARAM params[2] =
        { OSSL_PARAM_construct_utf8_string(key, name, 0),
          OSSL_PARAM_END };
````
Somehow it would have to figure out the "type" for the key, either by querying the "settable" params, or possibly by inferring it from the value.

You'd probably need a variant of `OSSL_Q_mac` that can *alternatively* take an `OSSL_PARAM` array.

Reply:
One can even come up with concise constructors such as 
`OSSL_PSTR("digest", "SHA256")` that would avoid the type and parsing issue.
These could allocate the parameter structure(s) on-the-fly, and the consuming function (such as `OSSL_*_mac()`) could take care of deallocating them.

Reply:
Unfortunately I think this is bound to cause problems. In general the code that allocated memory should be responsible for de-allocating it. De-allocating it as a side effect of some other function seems like an anti-pattern to me.

Reply:
You do know that this discussion is heading straight toward ideas like #14543, right?  I'm starting to wonder if it was a good idea to push that post 3.0.0, as we're now working on more complex kludges...

Reply:
> Unfortunately I think this is bound to cause problems. In general the code that allocated memory should be responsible for de-allocating it. De-allocating it as a side effect of some other function seems like an anti-pattern to me.

I agree on that in general, but here it would be a very narrow application domain, 
namely syntactic sugar for concise/convenient/flexible parameter provision,
where IMHO an exception to the general rule could be acceptable.

Reply:
> You'd probably need a variant of `OSSL_Q_mac` that can *alternatively* take an `OSSL_PARAM` array.

Isn't that called `NULL`?

Reply:
Yeah, one can have both flavors of crypto parameter lists as parameters of a single function, where any of them may be NULL.

Reply:
Here's a further pragmatic thought: how about
```
unsigned char *EVP_mac(OSSL_LIB_CTX *libctx, const char *mac_name, const char *propq,
                       const char *digest, const OSSL_PARAM *further_params,
                       const void *key, int keylen,
                       const unsigned char *data, size_t datalen,
                       unsigned char *out, size_t outsize, unsigned int *outlen)
```

Reply:
So a typical use would be
```
OSSL_mac(libctx, "HMAC", propq, "SHA256", NULL,
         key, keylen, data, datalen, out, outsize, outlen);
 ```

Reply:
Going with the idea from #14543, I raise you:

``` C
unsigned char *OSSL_prf(OSSL_LIB_CTX *libctx, const char *prf_name, const char *propq,
                       const OSSL_PARAM *params, const void *key, int keylen,
                       const unsigned char *data, size_t datalen,
                       unsigned char *out, size_t outsize, unsigned int *outlen)
```

Example usage:

``` C
OSSL_prf(libctx, "HMAC-SHA256", propq, NULL,
         key, keylen, data, datalen, out, outsize, outlen);
```

(oh, and if the provider has added an OID among the names, an OBJ_obj2txt() that results in either "HMAC-SHA256" or "1.2.840.113549.2.9" would work equally well)

Reply:
This PR already takes care of strongly simplifying the PBMAC calculation in `crmf_pbm.c`, currently like this:
```
(OSSL_HMAC(libctx, propq, mdname, basekey, bklen, msg, msglen,
           mac_res, EVP_MAX_MD_SIZE, &maclen)
```
Making use of my above proposal, this is going to become
```
(OSSL_mac(libctx, "HMAC", NULL, mdname, NULL, basekey, bklen, 
          msg, msglen, mac_res, EVP_MAX_MD_SIZE, &maclen)
```

Reply:
Ok to go for that?

Reply:
If you want a true comment on that particular piece of code, I find it deeply problematic that it hard-codes HMAC.  Yes, that's most of what `EVP_PBE_find()` currently supports, but still.

Reply:
I tried to fix the hard-coded "HMAC" but failed.
So I'd like to leave the `OSSL_prf()` topic for your care in #14543.

Reply:
Yet for other applications, ok to replace
```
 unsigned char *OSSL_Q_mac(OSSL_LIB_CTX *libctx, const char *mac_name,
                           const char *propq, const OSSL_PARAM *params,
                           const void *key, int keylen,
                           const unsigned char *data, size_t datalen,
                           unsigned char *out, size_t outsize,
                           unsigned int *outlen);
 unsigned char *EVP_MAC_calc(EVP_MAC *mac, const OSSL_PARAM *params,
                             const void *key, int keylen,
                             const unsigned char *data, size_t datalen,
                             unsigned char *out, size_t outsize,
                             unsigned int *outlen);
```
by
```
unsigned char *EVP_mac(OSSL_LIB_CTX *libctx, const char *name, const char *propq,
                       const char *digest, const OSSL_PARAM *params,
                       const void *key, int keylen,
                       const unsigned char *data, size_t datalen,
                       unsigned char *out, size_t outsize, unsigned int *outlen);
```
?

Reply:
For CMAC, do you see passing the cipher name as `digest`?

Reply:
Hmm, good point, then maybe call that parameter `subalg`?

Reply:
I'm confused.  Why is this hardcoding HMAC?  It's just a string isn't it?

For CMAC we'd want a `cipher` not a `digest`.  We could use `subalg`, which would work until someone comes along with an algorithm that uses more than one.  This is the case for at least one of the KDFs and is why it wasn't called `algorithm`.  BTW, some MACs don't require an algorithm at all (siphash).

I quite like Matt's suggestion of passing a string in and parsing it.  The property parser does almost exactly what we require.  Functions that convert params to/from a string given the gettable/settable arrays wouldn't be much of a stretch from what we've got already.

This isn't the first time we've hit all of this stuff.  The aforementioned KDFs for the `algorithm` parametert.  The HMAC DRBG allows any MAC to be used, not just HMACs.  It accept params and chooses the ones it needs to: `cipher` for a cipher and `digest` for a digest.  Even `mac` for the name of the MAC to use.

Reply:
> I'm confused. Why is this hardcoding HMAC? It's just a string isn't it?

Yes, and the point with the use in `OSSL_CRMF_pbm_new() in `crmf_pbm.c` is that although the mac alg should be entirely flexible and thus allow also non-HMAC, currently the main alg is fixed to "HMAC" (and in this sense hard-coded),
which likely is going to be fixed along with #14543.

> For CMAC we'd want a `cipher` not a `digest`. 

Sure, and `EVP_mac()` can cater for that automatically,
in a way that was already used (in a little more involved way) in `evp_test.c`.
I've tried it out in the current version of the PR, and it works fine:
```
    if (subalg != NULL) {
        const OSSL_PARAM *defined_params = EVP_MAC_settable_ctx_params(mac);
        const char *param_name = OSSL_MAC_PARAM_DIGEST;

        /*
         * The underlying algorithm may be a cipher or a digest.
         * We don't know which it is, but we can ask the MAC what it
         * should be and bet on that.
         */
        if (OSSL_PARAM_locate_const(defined_params, param_name) == NULL) {
            param_name = OSSL_MAC_PARAM_CIPHER;
            if (OSSL_PARAM_locate_const(defined_params, param_name) == NULL) {
                ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_INVALID_ARGUMENT);
                goto err;
            }
        }
        subalg_param[0] =
            OSSL_PARAM_construct_utf8_string(param_name, (char *)subalg, 0);
    }
```

>We could use `subalg`, which would work until someone comes along with an algorithm that uses more than one. This is the case for at least one of the KDFs and is why it wasn't called `algorithm`.

Even any such more gerneral situations can be supported easily by the 
proposed extra `const OSSL_PARAM *params` argument.

> BTW, some MACs don't require an algorithm at all (siphash).

Yeah, and also this is covered by allowing the `subalg` string to be NULL,
as given in the code I quoted above.

> I quite like Matt's suggestion of passing a string in and parsing it. The property parser does almost exactly what we require. Functions that convert params to/from a string given the gettable/settable arrays wouldn't be much of a stretch from what we've got already.

With the proposed solution, tentatively implemented here, 
this rather cumbersome and error-prone parsing is not needed.